rp2: Disable the LWIP tick timer when not needed.

projectgus opened the pull request:
### Summary

Closes #16181. Prevents lightsleep being woken up every 64ms to service LWIP timers, when:

1. No netif is up, and
2. No TCP sockets are active

The TCP socket check may not be strictly necessary, but without ticking the tcp timer they won't ever time out by themselves.

It means that it's possible to do:

```py
wlan.disconnect()
machine.lightsleep(5000)
wlan.connect(...)
```

... and get the full sleep time, but also possible to do 

```py
while some_condition():
    machine.lightsleep(5000)
```

... and Python will wake up every 64ms if Wi-Fi is active, but power consumption is still lower than it would be otherwise.

(Not massively lower, on RPI_PICO_W I measure 53mA idle at REPL with Wi-Fi associated, and 40mA in a `while True: machine.lightsleep(1000)` loop. There is probably potential to make this better by having lightsleep put the CYW43 driver into a lower power mode.)

*This work was funded through GitHub Sponsors.*

### Testing

**Important: This PR has only been tested cherry-picked onto the branch for #16454. It is marked as a draft until that PR is merged and it can be rebased.**

* Ran `multi_net` and `multi_wlan` tests on RPI_PICO_W and RPI_PICO2_W. Same failures as reported against master.

### Trade-offs and Alternatives

* There is some risk that some part of LWIP doesn't like being starved of timeouts for a long period after a network interface goes down. However, it looks like it will be OK - the main thing is TCP/IP socket timeouts and LWIP will keep ticking until these time out.
* Could make LWIP timer not a wake-up source, but this is quite difficult and does prevent using light sleep effectively with Wi-Fi.
* Could make LWIP timer wake-ups not return from lightsleep (i.e. the hardware wakes up, services LWIP, then goes back to sleep), but this is also quite fiddly to do (need to check the wake source each time).
* Semi-related: there is also a LWIP function `sys_timeouts_sleeptime()` which could be used to have the LWIP timer run at an exact time instead of at a constant tick rate. However, a number of LWIP "cyclic" timers run every 100ms right now so this change seems like it would be of limited benefit, it would add some code size, *and* the lwip tick callback is also used to poll the wiznet5k driver.

projectgus committed:
rp2: Disable the LWIP tick timer when not needed.

Prevents lightsleep being woken up every 64ms to service LWIP timers, when:

1. No netif is up, and
2. No TCP sockets are active

The TCP socket check may not be strictly necessary, but without ticking the
tcp timer they won't ever time out by themselves.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

madozu commented:
I did a lot of testing with the current master branch and pull requests #16454 and #17272 applied.

All the "early wakeup" scenarios I was experiencing when using machine.lightsleep() are solved --> **Great news, many thanks** to @projectgus for providing a good solution 😄 

So far, I did not encounter any negative side effects.

dpgeorge commented on ports/rp2/mpnetworkport.c:
```diff
@@ -147,7 +170,16 @@ void mod_network_lwip_init(void) {
+static void mp_network_netif_status_cb(struct netif *netif, netif_nsc_reason_t reason, const netif_ext_callback_args_t *args) {
+    // Start the network soft timer any time an interface comes up
+    if (reason == LWIP_NSC_LINK_CHANGED && args->link_changed.state) {
+        soft_timer_reinsert(&mp_network_soft_timer, LWIP_TICK_RATE_MS);
```
Maybe this needs a check `... && mp_network_soft_timer.mode == SOFT_TIMER_MODE_ONE_SHOT` so that it's only reinserted if it was disabled.

That would prevent pathological cases like an interface going up/down continuously and constantly pushing the timer out further and further.

projectgus replied:
Check added. Also needed to switch the `mode` here to avoid the pathological case (otherwise too-fast interface flapping can still starve the timeout).

projectgus commented:
Rebased, addressed review comments, re-ran `multi_net` and `multi_wifi` tests on RPI_PICO_W & RPI_PICO2_W with same results.

* On my local setup `multi_net/udp_data.py` fails intermittently to packet loss - but I have a mediocre wifi signal here.
*  `multi_net/tls_dtls_server_client.py` fails when the rp2 is instance 1.
* `multi_wlan` I realised has a bug when STA is rp2 and a wifi connection is already active. Will submit separate PR to fix.

dpgeorge reviewed (approved):
Tested on RPI_PICO_W and RPI_PICO2_W, running all possible tests:
- `tests/ports/rp2/rp2_lightsleep_thread.py` fails on both of them, but that's expected (and fails also on master)
- `tests/multi_net/udp_data_multi.py` fails when one of the boards is instance 0, and also fails on master.  I was sure this was working when I wrote the test so maybe it's down to local WLAN variability.

All other tests pass for me.
